File: apps/web/src/lib/appDownloads.ts
import { apiFetch } from "./api";

export const APP_DOWNLOAD_ANALYTICS = {
  viewed: "mywave_event_app_card_viewed",
  selected: "mywave_event_app_platform_selected",
  clicked: "mywave_event_app_download_clicked",
  succeeded: "mywave_event_app_download_succeeded",
  failed: "mywave_event_app_download_failed",
} as const;

export type AppDownloadArtifactId = "android" | "ios" | "source" | "documentation";

export type AppDownloadArtifactState =
  | "loading"
  | "available"
  | "unavailable"
  | "error"
  | "success";

export type AppDownloadArtifact = {
  id: AppDownloadArtifactId | string;
  label: string;
  platform: string;
  format: string;
  version: string;
  size: string | null;
  last_updated: string;
  action_label: string;
  requirements: string[];
  state: AppDownloadArtifactState | string;
  message: string;
};

export type AppDownloadManifest = {
  app: {
    id: string;
    name: string;
    short_description: string;
    features: string[];
    version: string;
    last_updated: string;
    platforms: string[];
    readiness: string;
    available_count: number;
  };
  artifacts: AppDownloadArtifact[];
  generated_at: string;
};

export type AppDownloadHandoff = {
  artifact_id: string;
  location: string;
  open_in_new_tab: boolean;
  message: string;
};

export function getAppDownloadManifest() {
  return apiFetch<AppDownloadManifest>("/api/v1/app-downloads/manifest");
}

export function getAppDownloadStatus(artifactId: string) {
  return apiFetch<AppDownloadArtifact>(
    `/api/v1/app-downloads/${encodeURIComponent(artifactId)}/status`,
  );
}

export function startAppDownloadHandoff(artifactId: string) {
  return apiFetch<AppDownloadHandoff>(
    `/api/v1/app-downloads/${encodeURIComponent(artifactId)}/handoff`,
    { method: "POST", body: "{}" },
  );
}

export function trackAppDownloadEvent(
  event: string,
  meta: Record<string, string | number | boolean | null | undefined> = {},
  context = "projects/checklist-org",
): void {
  const properties: Record<string, string | number | boolean> = { app_id: "mywave-event-app" };
  for (const [key, value] of Object.entries(meta)) {
    if (value === undefined || value === null) continue;
    properties[key] = value;
  }
  void apiFetch("/api/v1/analytics/events", {
    method: "POST",
    body: JSON.stringify({
      event,
      context,
      channel: "web",
      properties,
    }),
  }).catch(() => {
    /* analytics must not block the UI */
  });
}

export function formatFileSize(size: string | null | undefined): string {
  return size && size.trim() ? size : "Уточняется";
}

File: services/api/app/schemas/app_downloads.py
"""Schemas for the app download catalog and analytics ingest."""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class AppDownloadAppOut(BaseModel):
    id: str
    name: str
    short_description: str
    features: list[str]
    version: str
    last_updated: str
    platforms: list[str]
    readiness: str
    available_count: int


class AppDownloadArtifactOut(BaseModel):
    id: str
    label: str
    platform: str
    format: str
    version: str
    size: str | None = None
    last_updated: str
    action_label: str
    requirements: list[str]
    state: str
    message: str


class AppDownloadManifestOut(BaseModel):
    app: AppDownloadAppOut
    artifacts: list[AppDownloadArtifactOut]
    generated_at: str


class AppDownloadHandoffOut(BaseModel):
    artifact_id: str
    location: str
    open_in_new_tab: bool
    message: str


class AnalyticsEventIn(BaseModel):
    model_config = ConfigDict(extra="ignore")

    event: str = Field(min_length=1, max_length=80)
    context: str | None = Field(default=None, max_length=120)
    channel: str | None = Field(default="web", max_length=40)
    properties: dict[str, Any] | None = None
    meta: dict[str, Any] | None = None


class AnalyticsEventAccepted(BaseModel):
    accepted: bool = True
    event: str
